### PR TITLE
Remove uses of ClassTemplateSpecializationDecl::getTypeAsWritten

### LINKED ISSUE #142: LLVM 19: Build failures
The following needs to be ported to the latest LLVM tip:

```
In file included from /tmp/cvise/clang_delta/ReplaceDerivedClass.cpp:19:
/tmp/cvise/clang_delta/CommonRenameClassRewriteVisitor.h: In member function 'bool clang_delta_common_visitor::CommonRenameClassRewriteVisitor<T>::VisitClassTemplatePartialSpecializationDecl(clang::ClassTemplatePartialSpecializationDecl*)':
/tmp/cvise/clang_delta/CommonRenameClassRewriteVisitor.h:147:39: error: 'class clang::ClassTemplatePartialSpecializationDecl' has no member named 'getTypeAsWritten'
  147 |     const TypeSourceInfo *TyInfo = D->getTypeAsWritten();
      |                                       ^~~~~~~~~~~~~~~~
/tmp/cvise/clang_delta/RemoveNamespace.cpp: In member function 'bool RemoveNamespaceRewriteVisitor::VisitClassTemplatePartialSpecializationDecl(clang::ClassTemplatePartialSpecializationDecl*)':
/tmp/cvise/clang_delta/RemoveNamespace.cpp:442:39: error: 'class clang::ClassTemplatePartialSpecializationDecl' has no member named 'getTypeAsWritten'
  442 |     const TypeSourceInfo *TyInfo = D->getTypeAsWritten();
      |                                       ^~~~~~~~~~~~~~~~
```

Can @nickdesaulniers or @strimo378 help me, please?

### PR DESCRIPTION
**Pull request by sdkrystian**
Fixes #142

### CODE REVIEW

**Comment by marxin** [comment]
The suggested change looks fine! However, it seems the additional changes to the renaming pass will be needed in order to fix the 2 failures spotted with LLVM 19.

@sdkrystian Are you willing to work on that as part of this PR?

**Comment by sdkrystian** [comment]
@marxin I pushed a commit which should fix the failing tests

File: clang_delta/RemoveNamespace.cpp [hunk 1]
```diff
@@ -132,14 +129,24 @@ bool RemoveNamespaceASTVisitor::VisitNamespaceDecl(NamespaceDecl *ND)
...
-       E = TSD->bases_end(); I != E; ++I) {
-    TypeSourceInfo *TSI = (*I).getTypeSourceInfo();
-    TransAssert(TSI && "Bad TypeSourceInfo!");
-    TraverseTypeLoc(TSI->getTypeLoc());
+  const CXXRecordDecl *CXXRD = TSD->getSpecializedTemplate()->getTemplatedDecl();
+  std::string Name;
+  if (getNewName(CXXRD, Name)) {
+    SourceLocation LocStart = TSD->getLocation();
+    TransAssert(LocStart.isValid() && "Invalid Location!");
+    TheRewriter->ReplaceText(LocStart, CXXRD->getNameAsString().size(), Name);
```
Comment (marxin) [inline clang_delta/RemoveNamespace.cpp:140]:
```suggestion
    ConsumerInstance->TheRewriter.ReplaceText(LocStart, CXXRD->getNameAsString().size(), Name);
```

File: clang_delta/RemoveNamespace.cpp [hunk 2]
```diff
@@ -132,14 +129,24 @@ bool RemoveNamespaceASTVisitor::VisitNamespaceDecl(NamespaceDecl *ND)
...
     return true;
 
-  for (CXXRecordDecl::base_class_const_iterator I = TSD->bases_begin(),
-       E = TSD->bases_end(); I != E; ++I) {
-    TypeSourceInfo *TSI = (*I).getTypeSourceInfo();
-    TransAssert(TSI && "Bad TypeSourceInfo!");
-    TraverseTypeLoc(TSI->getTypeLoc());
+  const CXXRecordDecl *CXXRD = TSD->getSpecializedTemplate()->getTemplatedDecl();
+  std::string Name;
+  if (getNewName(CXXRD, Name)) {
```
Comment (marxin) [inline clang_delta/RemoveNamespace.cpp:137]:
```suggestion
  if (ConsumerInstance->getNewName(CXXRD, Name)) {
```

**Review by marxin — CHANGES_REQUESTED** [review]
I think the following changes are needed for a proper compilation.

File: clang_delta/RemoveNamespace.cpp (hunk 1 above)
Comment (sdkrystian) [reply clang_delta/RemoveNamespace.cpp:140]: Oops :P

**Comment by sdkrystian** [comment]
@marxin updated

**Comment by marxin** [comment]
The current LLVM master seems promising, but we will likely need a conditional build for the existing releases. What do you think 🤔?

**Comment by marxin** [comment]
The CI should be happy after 16a34b273966e32843b3f193590198e2a769b1b0.

**Pull request merged**